### PR TITLE
[IMP] attachment_s3: DB Operations

### PR DESCRIPTION
**Pull request by max3903**
#223

### CODE REVIEW

File: attachment_s3/__init__.py [hunk 1]
```diff
@@ -1,2 +1,10 @@
...
 from . import models
+from . import controllers
+
+from odoo import api, SUPERUSER_ID
+
+
+def _auto_create_bucket(cr, registry):
+    # create the S3 bucket after module installed
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    env["ir.attachment"]._get_s3_bucket()
```
Comment (max3903) [inline attachment_s3/__init__.py]: @hiren-serpentcs You also need to set the config parameter and force the storage like here:
https://github.com/ursais/odoo-template/blob/14.0/odoo/songs/setup/base.py#L51

File: attachment_s3/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+_logger = logging.getLogger(__name__)
+
+
+class Database(Database):
+    @http.route()
+    def drop(self, master_pwd, name):
+        response = super(Database, self).drop(master_pwd, name)
+        try:
+            bucket = request.env["ir.attachment"]._get_s3_bucket()
+            bucket.objects.all().delete()
```
Comment (max3903) [inline attachment_s3/controllers/main.py]: @hiren-serpentcs Does this also delete the bucket itself?

**Comment by max3903** [comment]
Ping @jjscarafia

**Comment by jjscarafia** [comment]
Hi @max3903 
Thanks for the ping.
IMHO, deleting the bucket when deleting the db should be optional. Not sure how, may be with an environment key? or a conf parameter? In my use case it would be very dangerous to delete the bucket.

my 2 cents

File: attachment_s3/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            bucket = request.env["ir.attachment"]._get_s3_bucket()
+            sql = ("""
+                UPDATE ir_attachment AS t SET store_fname = s.store_fname FROM (
+                    SELECT
+                        id,
+                        REPLACE(store_fname, '/*production-master*/', '%s')
+                    AS store_fname FROM ir_attachment WHERE db_datas is NULL)
+                AS s(id,store_fname) where t.id = s.id;
+            """ (bucket.name,))
+            request.env.cr.execute(sql)
```
Comment (max3903) [inline attachment_s3/controllers/main.py]: @hiren-serpentcs The database is already dropped.

File: attachment_s3/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            s3_resource = request.env["ir.attachment"]._get_s3_client()
+            bucket_to_copy = request.env["ir.attachment"]._get_s3_bucket()
+            new_bucket_name = request.env["ir.attachment"]._get_s3_bucket(new_name)
+            new_bucket_name = new_bucket_name.name
+            bucket_to_copy = bucket_to_copy.name
+            sql = ("""
+                UPDATE ir_attachment AS t SET store_fname = s.store_fname FROM (
+                    SELECT
+                        id,
+                        REPLACE(store_fname, '/*production-master*/', '%s')
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
                        REPLACE(store_fname, '/*%s*/', '%s')
```

File: attachment_s3/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                    AS store_fname FROM ir_attachment WHERE db_datas is NULL)
+                AS s(id,store_fname) where t.id = s.id;
+            """ (new_bucket_name,))
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
            """ (bucket_to_copy, new_bucket_name))
```

File: attachment_s3/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            request.env.cr.execute(sql)
+            for key in s3_resource.list_objects(Bucket=bucket_to_copy)["Contents"]:
+                files = key["Key"]
+                copy_source = {"Bucket": bucket_to_copy, "Key": files}
+                s3_resource.meta.client.copy(copy_source, new_bucket_name, files)
+
+            return response
+        except exceptions.UserError:
+            _logger.exception("error reading attachment from object storage")
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
            _logger.exception("Error reading attachments from object storage.")
```

File: attachment_s3/controllers/main.py (hunk 3 above)
Comment (max3903) [reply attachment_s3/controllers/main.py]: @hiren-serpentcs We don't need this. The database is being dropped, why would we update it?

File: attachment_s3/__init__.py [hunk 7]
```diff
@@ -1,2 +1,11 @@
...
```
Comment (max3903) [inline attachment_s3/__init__.py]:
```suggestion
    env["ir.attachment"]._get_s3_bucket()
    env["ir.config_parameter"].create({"key": "ir_attachment.location", "value": "s3"})
```

File: attachment_s3/controllers/main.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            bucket = request.env["ir.attachment"]._get_s3_bucket()
+            sql = ("""
+                UPDATE ir_attachment AS t SET store_fname = s.store_fname FROM (
+                    SELECT
+                        id,
+                        REPLACE(store_fname, '/*production-master*/', '%s')
+                    AS store_fname FROM ir_attachment WHERE db_datas is NULL)
+                AS s(id,store_fname) where t.id = s.id;
+            """ (bucket.name,))
+            request.env.cr.execute(sql)
```
Comment (max3903) [inline attachment_s3/controllers/main.py:25]:
```suggestion
```

Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
```

File: attachment_s3/controllers/main.py [hunk 9]
```diff
@@ -12,6 +12,9 @@
 class Database(Database):
     @http.route()
     def drop(self, master_pwd, name):
+        delete_on_drop = os.environ.get("AWS_DELETE_ON_DBDROP")
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
        delete_on_drop = os.environ.get("AWS_DELETE_ON_DBDROP", False)
```

File: attachment_s3/models/ir_attachment.py [hunk 10]
```diff
@@ -53,12 +54,13 @@ def _get_s3_client(self):
             params["endpoint_url"] = host
         if region_name:
             params["region_name"] = region_name
-        if not (access_key and secret_key):
+        if not (access_key and secret_key and delete_on_drop):
```
Comment (max3903) [inline attachment_s3/models/ir_attachment.py]:
```suggestion
        if not (access_key and secret_key):
```

File: attachment_s3/models/ir_attachment.py [hunk 11]
```diff
@@ -53,12 +54,13 @@ def _get_s3_client(self):
...
             msg = _(
                 "If you want to read from the S3 bucket, the following "
                 "environment variables must be set:\n"
                 "* AWS_ACCESS_KEY_ID\n"
                 "* AWS_SECRET_ACCESS_KEY\n"
+                "* AWS_DELETE_ON_DBDROP\n"
```
Comment (max3903) [inline attachment_s3/models/ir_attachment.py]:
```suggestion
```

File: attachment_s3/controllers/main.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            """ (bucket_to_copy, new_bucket_name,))
+            request.env.cr.execute(sql)
+            for key in s3_resource.list_objects(Bucket=bucket_to_copy)["Contents"]:
+                files = key["Key"]
+                copy_source = {"Bucket": bucket_to_copy, "Key": files}
+                s3_resource.meta.client.copy(copy_source, new_bucket_name, files)
+
+            return response
+        except exceptions.UserError:
+            _logger.exception("Error reading attachments from object storage.")
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
            _logger.exception("Error writing attachments to object storage.")
```

File: attachment_s3/controllers/main.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        response = super(Database, self).restore(master_pwd, backup_file, name, copy)
+        try:
+            request.env["ir.attachment"]._get_s3_bucket(name)
+            request.env["ir.config_parameter"].create(
+                {"key": "ir_attachment.location", "value": "s3"}
+            )
+            request.env["ir.attachment"].force_storage()
+            return response
+        except exceptions.UserError:
+            _logger.exception("Error reading attachments from object storage.")
```
Comment (max3903) [inline attachment_s3/controllers/main.py]:
```suggestion
            _logger.exception("Error writing attachments to object storage.")
```

**Review by nbessi — CHANGES_REQUESTED** [review]
I'm not fan of this approach. That should be the role of an operator also that pose a lot of compatibility question and might be incompatible with the pod affinity/identity  philosophy promoted by Cloud providers. I would prefer to have that as a separated module to be honnest

**Pull request closed without merging**